api/v1alpha2/awsmachine_conversion.go

```diff
@@ -19,23 +19,61 @@ package v1alpha2
 import (
 	apiconversion "k8s.io/apimachinery/pkg/conversion"
 	infrav1alpha3 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+	utilconversion "sigs.k8s.io/cluster-api/util/conversion"
 	"sigs.k8s.io/controller-runtime/pkg/conversion"
 )
 
 // ConvertTo converts this AWSMachine to the Hub version (v1alpha3).
 func (src *AWSMachine) ConvertTo(dstRaw conversion.Hub) error { // nolint
 	dst := dstRaw.(*infrav1alpha3.AWSMachine)
+
 	if err := Convert_v1alpha2_AWSMachine_To_v1alpha3_AWSMachine(src, dst, nil); err != nil {
 		return err
 	}
 
+	// Manually restore data from annotations
+	restored := &infrav1alpha3.AWSMachine{}
+	if ok, err := utilconversion.UnmarshalData(src, restored); err != nil || !ok {
+		return err
+	}
+
+	restoreAWSMachineSpec(&restored.Spec, &dst.Spec)
+
 	return nil
 }
 
+func restoreAWSMachineSpec(restored *infrav1alpha3.AWSMachineSpec, dst *infrav1alpha3.AWSMachineSpec) {
+	dst.ImageLookupBaseOS = restored.ImageLookupBaseOS
+
+	// Restore CloudInit according to the following rules:
+	// 1. If no CloudInit data exists on converted object, then replace entirely with re-hydrated version.
+	// 2. If CloudInit data exists on both the new converted object  and the re-hydrated version, we
+	// preferentially take the SecretARN if it exists on the converted object.
+	// 3. If a Secret ARN exists on the re-hydrated object, but not the converted one, take the re-hydrated
+	// value of the Secret ARN
```

Comment:
I need to think through this some more (doesn't have to block merging) - I'm not convinced we need to handle a case where we have a v1alpha2 object with CloudInit both in the spec and in the annotation.

Reply:
Admittedly a super-narrow use case: transferring from a v1alpha3 cluster to an old v1alpha2 cluster and back again.